File: java/abc206/C.java
import java.util.*;

public class C {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int N = sc.nextInt();
    long[] As = new long[(int) N];
    Arrays.setAll(As, i -> sc.nextInt());

    Map<Long, Long> hm = new HashMap<>();
    long cnt = 0;

    for (int i = 0; i < N; i++) {
      long target = As[N - i - 1];
      long target_cnt = hm.getOrDefault(target, 0L);
      cnt += (i - target_cnt);
      hm.put(target, target_cnt + 1);
    }

    System.out.println(cnt);
  }
}

File: rust/abc206/src/c.rs
use proconio::input;
use std::collections::HashMap;
use std::io::{stdout, BufWriter, Write};

pub fn main() {
  let out = stdout();
  let mut out = BufWriter::new(out.lock());
  input! {n: usize, a: [usize; n]}
  let mut hs: HashMap<usize, usize> = HashMap::with_capacity(n);
  let mut cnt = 0;
  for i in 0..n {
    let target = a[n - i - 1];
    let target_cnt = hs.entry(target).or_insert(0);
    cnt += i - *target_cnt;
    *target_cnt += 1;
  }
  writeln!(out, "{}", cnt).unwrap();
}
